src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs

```diff
@@ -30,179 +30,171 @@ protected MessageBody(Frame context)
-                }
-
-                Task destinationTask;
                 try
                 {
-                    destinationTask = destination.WriteAsync(segment.Array, segment.Offset, segment.Count, cancellationToken);
-                }
-                catch
-                {
-                    ConsumedBytes(segment.Count);
-                    throw;
+                    if (!readableBuffer.IsEmpty)
+                    {
+                        var actual = Math.Min(readableBuffer.Length, buffer.Count);
+                        var slice = readableBuffer.Slice(0, actual);
+                        consumed = readableBuffer.Move(readableBuffer.Start, actual);
+                        slice.CopyTo(buffer);
+                        return actual;
+                    }
+                    else if (result.IsCompleted)
+                    {
+                        return 0;
+                    }
                 }
-
-                if (!destinationTask.IsCompleted)
+                finally
                 {
-                    return CopyToAsyncDestinationAwaited(destinationTask, segment.Count, destination, cancellationToken);
+                    _context.RequestBodyPipe.Reader.Advance(consumed);
                 }
-
-                ConsumedBytes(segment.Count);
-
-                // Surface errors if necessary
-                destinationTask.GetAwaiter().GetResult();
-
-                peekTask = PeekAsync(cancellationToken);
             }
-
-            TryProduceContinue();
-
-            return CopyToAsyncPeekAwaited(peekTask, destination, cancellationToken);
         }
 
-        private async Task CopyToAsyncPeekAwaited(
-            ValueTask<ArraySegment<byte>> peekTask,
-            Stream destination,
-            CancellationToken cancellationToken = default(CancellationToken))
+        public virtual async Task CopyToAsync(Stream destination, CancellationToken cancellationToken = default(CancellationToken))
         {
             while (true)
             {
-                var segment = await peekTask;
-
-                if (segment.Count == 0)
-                {
-                    return;
-                }
+                var result = await _context.RequestBodyPipe.Reader.ReadAsync();
+                var readableBuffer = result.Buffer;
+                var consumed = readableBuffer.End;
 
                 try
                 {
-                    await destination.WriteAsync(segment.Array, segment.Offset, segment.Count, cancellationToken);
+                    if (!readableBuffer.IsEmpty)
+                    {
+                        foreach (var memory in readableBuffer)
+                        {
+                            var array = memory.GetArray();
+                            await destination.WriteAsync(array.Array, array.Offset, array.Count, cancellationToken);
+                        }
+                    }
+                    else if (result.IsCompleted || result.IsCancelled)
```

Comment:
I think we should throw `TaskCanceledException` on `IsCancelled`. Any reason not to?